dissect/target/containers/fortifw.py

```diff
@@ -0,0 +1,206 @@
+    buf = fobj.read(32)
+    fobj.seek(pos)
+
+    if pos % 512 == 0:
+        xor_char = 0xFF
+    else:
+        fobj.seek(pos - 1)
+        xor_char = ord(fobj.read(1))
+
+    for i, k_char in enumerate(buf):
+        idx = (i + pos) & 0x1F
+        key.append((xor_char ^ k_char ^ idx) & 0xFF)
+        xor_char = k_char
+
+    # align xor key
+    koffset = 32 - (pos & 0x1F)
+    key = islice(cycle(key), koffset, koffset + 32)
+    return bytes(key)
+
+
+class FortiFirmwareFile:
+    """Fortinet firmware file, handles transparant decompression and deobfuscation of the firmware file."""
+
+    def __init__(self, fobj: io.BytesIO) -> None:
+        self.fh = fobj
+        self.size = None
+
+        # Check if the file is gzipped
+        self.is_gzipped = False
+        self.fh.seek(0)
+        header = self.fh.read(4)
+        if header.startswith(b"\x1f\x8b"):
+            self.is_gzipped = True
+
+            # Find the extra metadata behind the gzip compressed data
+            # as a bonus we can also calculate the size of the firmware here
+            dec = zlib.decompressobj(wbits=16 + zlib.MAX_WBITS)
+            self.fh.seek(0)
+            self.size = 0
+            while True:
+                data = self.fh.read(io.DEFAULT_BUFFER_SIZE)
+                if not data:
+                    break
+                d = dec.decompress(dec.unconsumed_tail + data)
+                self.size += len(d)
+
+            # Ignore the trailer data of the gzip file if we have any
+            if dec.unused_data:
+                self.fh.seek(-len(dec.unused_data), os.SEEK_END)
+                self.trailer_offset = self.fh.tell()
+                self.trailer_data = self.fh.read()
+                logger.info("Found trailer offset: %d, data: %r", self.trailer_offset, self.trailer_data)
+                self.fh = RangeStream(self.fh, 0, self.trailer_offset)
+
+            self.fh.seek(0)
+            self.fh = gzip.GzipFile(fileobj=self.fh)
+
+        # Find the xor key based on known offsets where the firmware should decode to zero bytes
+        for zero_offset in (0x30, 0x40, 0x400):
+            self.fh.seek(zero_offset)
+            xor_key = find_xor_key(self.fh)
+            if xor_key.isalnum():
+                self.xor_key = xor_key
+                logger.info("Found key %r @ offset %s", self.xor_key, zero_offset)
+                break
+        else:
+            self.xor_key = None
+            logger.info("No xor key found")
+
+        # Determine the size of the firmware file if we didn't calculate it yet
+        if self.size is None:
+            self.fh.seek(0, io.SEEK_END)
+            self.size = self.fh.tell()
+
+        logger.info("firmware size: %s", self.size)
+        logger.info("key: %r", self.xor_key)
+        logger.info("gzipped: %s", self.is_gzipped)
+        self.fh.seek(0)
+
+    def seek(self, offset, whence=io.SEEK_SET):
+        return self.fh.seek(offset, whence)
+
+    def read(self, n=-1):
```

Comment:
Have you tried turning this class into a subclass of `AlignedStream`?

Reply:
what benefit does that give?

Reply:
Some caching, read offset is always aligned, real compatibility with `io.RawIOBase`.

Reply:
Fixed, AlignedStream is awesome